src/iqm/cirq_iqm/devices/iqm_device.py

```diff
@@ -77,17 +84,35 @@ def get_qubit(self, index: int) -> cirq.Qid:
 
     def check_qubit_connectivity(self, operation: cirq.Operation) -> None:
         """Raises a ValueError if operation acts on qubits that are not connected."""
-        if len(operation.qubits) >= 2 and not isinstance(operation.gate, ops.MeasurementGate):
-            if operation.qubits not in self._metadata.nx_graph.edges:
-                raise ValueError(f'Unsupported qubit connectivity required for {operation!r}')
+        if len(operation.qubits) >= 2 and not self.has_valid_operation_targets(operation):
+            raise ValueError(f'Unsupported qubit connectivity required for {operation!r}')
 
     def is_native_operation(self, op: cirq.Operation) -> bool:
         """Predicate, True iff the given operation is considered native for the architecture."""
-        return (
+        check = (
             isinstance(op, (ops.GateOperation, ops.TaggedOperation))
             and (op.gate is not None)
             and (op.gate in self._metadata.gateset)
         )
+        if check and isinstance(op.gate, ops.CZPowGate):
+            return op.gate.exponent == 1
+        return check
+
+    def has_valid_operation_targets(self, op: cirq.Operation) -> bool:
+        """Predicate, True iff the given operation is native and it=s targets are valid."""
+        matched_support = [
+            (g, qbs)
+            for g, qbs in self.supported_operations.items()
+            if op.gate is not None and op.gate in cirq.GateFamily(g)
+        ]
+        if len(matched_support) > 0:
+            gf, valid_targets = matched_support[0]
+            if gf == cirq.MeasurementGate:  # Measurements can be done on any available qubits
+                return all(q in [q for qb in valid_targets for q in qb] for q in op.qubits)
```

Comment:
This runs the inner comprehension anew for each q. How about first collecting valid targets as a set and then using that in the outer comprehension?